src/main/java/com/louis/HashTable.java

```diff
@@ -0,0 +1,219 @@
+ */
+public class HashTable<K, V> {
+
+    private static final int MIN_CAPACITY = 32;
+    private static final double RESIZE_LOAD = 0.5;
+    private static final double MAX_LOAD = 0.75;
+    private static final double MIN_LOAD = 0.25;
+
+    private HashEntry[] table;
+    private int currentSize;
+    private int currentCapacity;
+
+
+    /**
+     * Creates a new HashTable.
+     */
+    @SuppressWarnings("unchecked")
+    public HashTable() {
+        currentSize = 0;
+        currentCapacity = MIN_CAPACITY;
+        table = (HashEntry[]) Array.newInstance(HashEntry.class, currentCapacity);
+    }
+
+    /**
+     * Defines a HashTable entry.
+     * Similar to java.util.AbstractMap.SimpleEntry.
+     */
+    private class HashEntry {
+
+        private final K key;
+        private final V value;
+
+        /**
+         * Creates a new HashEntry.
+         */
+        HashEntry(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        /**
+         * Returns the key.
+         * @return Returns the entries' key.
+         */
+        public K getKey() {
+            return key;
+        }
+
+        /**
+         * Returns the value.
+         * @return Returns the entries' value.
+         */
+        public V getValue() {
+            return value;
+        }
+    }
+
+    /**
+     * Returns the value associated with a key.
+     *
+     * @param key is the specified key.
+     * @return Returns the value associated with the key; null if no such value.
+     * @throws IllegalArgumentException is thrown if key is null.
+     */
+    public synchronized V get(K key) {
+        Preconditions.checkArgument(key != null, "first argument to get(K key) is null");
+        // Finds the key.
+        int index = hash(key) % currentCapacity;
+        while (!key.equals(table[index].getKey())) {
+            index = (index + 1) % currentCapacity;
+        }
+        return table[index] != null ? table[index].getValue() : null;
+    }
+
+    /**
+     * Adds a key and its value to the table.
+     *
+     * @param key is the key to be added.
+     * @param value is the value for this key.
+     * @throws IllegalArgumentException is thrown if key or value is null.
+     */
+    public synchronized void put(K key, V value) {
+        Preconditions.checkArgument(key != null, "first argument to put(K key, ...) is null");
+        int index = hash(key) % currentCapacity;
+        Preconditions.checkArgument(value != null, "second argument to put(..., V Value) is null");
+        while (table[index] != null && !key.equals(table[index].getKey())) {
+            index = (index + 1) % currentCapacity;
+        }
+
+        // Increases the size if a new key is being put in.
+        if (table[index] == null) {
+            currentSize++;
+        }
+        table[index] = new HashEntry(key, value);
+
+        resizeIfRequired();
+    }
+
+    /**
+     * Hashs a key.
```

Comment:
Hashs a key in order to calculate the desired position of the corresponding value within the underlying table.